### PR TITLE
Don't render when code ahead

### PR DESCRIPTION
**Pull request by Rheeseyb**
**Problem:**
When making code changes, we'll cause the canvas to re-render twice: once when the code is changed; and then a second time once the parsed model has caught up.

**Fix:**
Specifically check if any code is ahead before triggering a canvas render. To do this I used the existing `shouldRerender` in `editor.tsx`, but realised that was actually a bit of a misnomer, and was gating logic that it shouldn't be. So I have explicitly used the relevant boolean checks around the individual parts of the code called after a dispatch.

Note that I have completely removed the check around `editorDispatchClosingOut`. I'm very much certain that we should never have been gating that with a check that something had changed, as the internal logic of that function is already checking that something actually changed before e.g. triggering a save.

My only concern with this was that if you are typing fast enough it would prevent the canvas from rendering until you paused briefly, because we don't change the revisions state following a parser-printer update if the update is stale here: <https://github.com/concrete-utopia/utopia/blob/67abcdd1d9fa11c0bf42be255cf655bba45a5a38/editor/src/core/shared/parser-projectcontents-utils.ts#L100-L109>
However, in practice, even on the hydrogen project this never happened, because the IndexedDB polling part which forms the code-editor-to-utopia communication channel is slower than parsing the code. So I then added a 600ms pause in the parsing to force the code updates to come through faster, which meant updating the canvas only happened when you stopped typing, and even in that forced case it actually didn't feel bad.

**Manual Tests:**
I hereby swear that:
- [x] I opened a hydrogen project and it loaded
- [x] I could navigate to various routes in Preview mode

### CODE REVIEW

File: editor/src/templates/editor.tsx [hunk 1]
```diff
@@ -450,7 +451,12 @@ export class Editor {
       const reactRouterErrorPreviouslyLogged = hasReactRouterErrorBeenLogged()
 
       const runDomWalker = shouldRunDOMWalker(dispatchedActions, dispatchResult)
-      const shouldRerender = !dispatchResult.nothingChanged || runDomWalker
```
Comment (Rheeseyb) [inline editor/src/templates/editor.tsx:453]: I noticed that `shouldRerender` was actually very badly named here, and so changed some of the gating of the code below accordingly. It's worth expanding the diff below to understand what is actually happening now

File: editor/src/templates/editor.tsx [hunk 2]
```diff
@@ -467,89 +473,90 @@ export class Editor {
             })
           })
         })
+      }
```
[the remark this answers is not in the document]
Comment (Rheeseyb) [reply editor/src/templates/editor.tsx:476]: `shouldRerender` was a bit of a misnomer, especially as it was gating logic that should run regardless of if that canvas rendered (e.g. the `editorDispatchClosingOut` at the end). There are cases where it makes sense to run the dom-walker without a canvas render (e.g. navigating to a different route). But that being said, the below condition should actually be `somethingChanged || runDomWalker` so I'll fix that.

[the remark this answers is not in the document]
Comment (Rheeseyb) [reply editor/src/templates/editor.tsx:476]: I updated it to `shouldRerender || runDomWalker` in the end

[the remark this answers is not in the document]
Comment (balazsbajorics) [reply editor/src/templates/editor.tsx:476]: hmmm, but shouldRerender is already `const shouldRerender =
        (somethingChanged || runDomWalker) &&
        !anyCodeAhead(dispatchResult.unpatchedEditor.projectContents)`, aren't you defeating the anyCodeAhead check then?

[the remark this answers is not in the document]
Comment (Rheeseyb) [reply editor/src/templates/editor.tsx:476]: The dom walker should run whenever the canvas is rendered, hence the `shouldRerender` check, **but** it should also run when we explicitly ask it to (e.g. when the editor is in play mode and the user has navigated to a different route), hence the `|| runDomWalker`.

**Review by balazsbajorics — APPROVED** [review]

**Review by ruggi — APPROVED** [review]

**Review by seanparsons — APPROVED** [review]

**Pull request merged**